packages/graph/hash_graph/lib/graph/src/api/rest/entity.rs

```diff
@@ -0,0 +1,189 @@
+//! Web routes for CRU operations on entities.
+
+use axum::{
+    extract::Path,
+    http::StatusCode,
+    response::IntoResponse,
+    routing::{get, post},
+    Extension, Json, Router,
+};
+use serde::{Deserialize, Serialize};
+use utoipa::{Component, OpenApi};
+
+use crate::{
+    api::rest::api_resource::RoutedResource,
+    knowledge::{Entity, EntityId},
+    ontology::{types::uri::VersionedUri, AccountId},
+    store::{
+        error::{EntityDoesNotExist, QueryError},
+        Store,
+    },
+};
+
+#[derive(Component, Serialize)]
+struct QualifiedEntity {
+    entity_id: EntityId,
+    entity: Entity,
+}
+
+#[derive(OpenApi)]
+#[openapi(
+    handlers(
+        create_entity,
+        get_entity,
+        update_entity
+    ),
+    components(CreateEntityRequest, UpdateEntityRequest, EntityId, QualifiedEntity),
+    tags(
+        (name = "Entity", description = "entity management API")
+    )
+)]
+pub struct EntityResource;
+
+impl RoutedResource for EntityResource {
+    /// Create routes for interacting with entities.
+    fn routes<S: Store>() -> Router {
+        // TODO: The URL format here is preliminary and will have to change.
+        Router::new().nest(
+            "/entity",
+            Router::new()
+                .route("/", post(create_entity::<S>).put(update_entity::<S>))
+                .route("/:entity_id", get(get_entity::<S>)),
+        )
+    }
+}
+
+#[derive(Serialize, Deserialize, Component)]
+struct CreateEntityRequest {
+    #[component(value_type = Any)]
+    entity: Entity,
+    entity_type_uri: VersionedUri,
+    account_id: AccountId,
+}
+
+#[utoipa::path(
+    post,
+    path = "/entity",
+    request_body = CreateEntityRequest,
+    tag = "Entity",
+    responses(
+      (status = 201, content_type = "application/json", description = "entity created successfully", body = QualifiedEntity),
+      (status = 422, content_type = "text/plain", description = "Provided request body is invalid"),
+
+      (status = 404, description = "Entity Type URI was not found"),
+      (status = 500, description = "Datastore error occurred"),
+    ),
+    request_body = CreateEntityRequest,
+)]
+async fn create_entity<S: Store>(
+    body: Json<CreateEntityRequest>,
+    datastore: Extension<S>,
+) -> Result<Json<QualifiedEntity>, StatusCode> {
+    let Json(CreateEntityRequest {
+        entity,
+        entity_type_uri,
+        account_id,
+    }) = body;
+    let Extension(datastore) = datastore;
```

Comment:
just bringing it into attention that we still use `datastore` for the params everywhere. Module restructure should probably drive-by update all these